### PR TITLE
achievements board

### PR DESCRIPTION
**Pull request by kostmo**
Towards #797.

This PR implements defines the achievements schema, defines a few example achievements, and implements saving and loading.

Main menu entry:
Screenshot from 2022-10-26 23-33-47

Achievements screen:
image

### CODE REVIEW

**Comment by xsebek** [comment]
How about adding the first achievements for completing some set of scenarios? 🙂 

We do not need to store new data for it, except for which scenarios are needed for the achievements.

@kostmo you could hardcode it in the app for now or add a `scenario-achievements.yaml` file if you want.

I think that would make this actually useful and also populate the Achievements view with some entries for a relatively little cost.

But I leave that to you, we had empty panels before, so we can merge it even as a mock-up. 👍

File: src/Swarm/TUI/Model/Achievement/Description.hs
Comment (byorgey) [inline src/Swarm/TUI/Model/Achievement/Description.hs:7]: Would there be any benefit to having all this information stored in an external `.yaml` file?  I know we still need to have a data type enumerating the possible achievements, since the code for deciding when we have obtained them is arbitrary and can't live in an external data file.  But the titles, flavor text, explanation, etc. could all live in a data file.

File: src/Swarm/TUI/Model/UI.hs
Comment (byorgey) [inline src/Swarm/TUI/Model/UI.hs:173]: This comment seems like a copy-pasto.

File: src/Swarm/TUI/Model/UI.hs
Comment (byorgey) [inline src/Swarm/TUI/Model/UI.hs:284]:
```suggestion
      , _uiAchievements = M.fromList $ map (view achievement &&& id) achievements
```

I find this clearer, but YMMV.  Feel free to use or ignore this suggestion as you wish.

File: src/Swarm/TUI/View/Achievement.hs
Comment (byorgey) [inline src/Swarm/TUI/View/Achievement.hs:32]: This wastes a lot of space and doesn't look very good on bigger terminals:
achievements-menu

Perhaps there should be some proportional padding to the left and right so the menu and achievement details don't extend all the way to the edges of the window?

File: src/Swarm/TUI/View/Achievement.hs
Comment (byorgey) [inline src/Swarm/TUI/View/Achievement.hs:59]: Since we store the time at which the achievement was attained, it might be nice to show that information somewhere.

File: src/Swarm/TUI/Controller.hs
Comment (byorgey) [inline src/Swarm/TUI/Controller.hs:162]: This is the only achievement out of the four that is actually achievable at the moment, right?  Or did I miss something?

File: src/Swarm/TUI/Model/StateUpdate.hs
Comment (byorgey) [inline src/Swarm/TUI/Model/StateUpdate.hs:124]: When will deferred achievements actually be saved?  Upon winning a scenario?  I don't think I saw any calls to `saveAchievementsInfo` anywhere else.  What kind of achievements do you envision being deferred vs immediate?

**Review by byorgey — COMMENTED** [review]
This is really nice so far!

File: src/Swarm/TUI/Model/Achievement/Description.hs
Comment (kostmo) [reply src/Swarm/TUI/Model/Achievement/Description.hs:7]: > we still need to have a data type enumerating the possible achievements

One of the members of this enumeration's sum type could have a string field, which makes it open-ended for achievements defined in config files.  Which raises the question, should all of the achievements be stored in config?  At the moment I think I prefer a hybrid: define as many achievements as feasible in code to obtain exhaustiveness-checking from the compiler, and let the rest be in `.yaml`.

Comment (byorgey) [reply src/Swarm/TUI/Model/Achievement/Description.hs:7]: Makes sense, although if we allow achievements to be defined completely in a `.yaml` file, wouldn't we need some sort of little DSL to describe what the achievement is for?  Or are you imagining that achievements could be defined via some Swarm code just like goal conditions?

Comment (kostmo) [reply src/Swarm/TUI/Model/Achievement/Description.hs:7]: Yes, they would be specially-annotated goals.  #795 is probably a prerequisite.

File: src/Swarm/TUI/Model/StateUpdate.hs
Comment (kostmo) [reply src/Swarm/TUI/Model/StateUpdate.hs:124]: I added this speculatively, but I suppose I'll remove it for now.

File: src/Swarm/TUI/View/Achievement.hs
Comment (kostmo) [reply src/Swarm/TUI/View/Achievement.hs]: Here's the new look:

| Narrow | Wide |
| --- | --- |
| Screenshot from 2022-12-17 14-53-55 | Screenshot from 2022-12-17 14-49-16 |

File: src/Swarm/TUI/Controller.hs
Comment (kostmo) [reply src/Swarm/TUI/Controller.hs]: At this moment there are two achievements implemented out of seven defined:

* "About Time!"
* "Call of the Void"

The latter demonstrates propagation of the achievement from within the GameState, out to the UI and persistence layers.

File: src/Swarm/Game/Step.hs
Comment (byorgey) [inline src/Swarm/Game/Step.hs:758]: Note this is currently only possible in creative mode, since it is not possible to craft a `detonator`.  That's fine though.

**Review by byorgey — APPROVED** [review]
This is looking great!  I'll leave it up to you whether you want to merge it as-is and incrementally work on adding more achievements, or wait until more has been implemented.

**Pull request merged**
